crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1129,6 +1246,73 @@ impl AuthorityPerEpochStore {
         Ok(())
     }
 
+    fn defer_transactions(
+        &self,
+        batch: &mut DBBatch,
+        key: DeferralKey,
+        transactions: Vec<VerifiedSequencedConsensusTransaction>,
+    ) -> SuiResult {
+        batch.insert_batch(
+            &self.tables.deferred_transactions,
+            std::iter::once((key, transactions)),
+        )?;
+        Ok(())
+    }
+
+    fn load_deferred_transactions_for_randomness_round(
+        &self,
+        batch: &mut DBBatch,
+        randomness_round: u64,
+    ) -> SuiResult<Vec<VerifiedSequencedConsensusTransaction>> {
+        let (min, max) = DeferralKey::range_for_randomness_round(randomness_round);
+        self.load_deferred_transactions(batch, min, max)
+    }
+
+    fn load_deferred_transactions_for_consensus_round(
+        &self,
+        batch: &mut DBBatch,
+        consensus_round: u64,
+    ) -> SuiResult<Vec<VerifiedSequencedConsensusTransaction>> {
+        let (min, max) = DeferralKey::range_for_consensus_round(consensus_round);
+        self.load_deferred_transactions(batch, min, max)
+    }
+
+    // factoring of the above
+    fn load_deferred_transactions(
+        &self,
+        batch: &mut DBBatch,
+        min: DeferralKey,
+        max: DeferralKey,
+    ) -> SuiResult<Vec<VerifiedSequencedConsensusTransaction>> {
+        let txns: Vec<_> = self
+            .tables
+            .deferred_transactions
+            .iter_with_bounds(Some(min), Some(max))
+            .flat_map(|(_, txns)| txns)
+            .collect();
+
+        // verify that there are no duplicates - should be impossible due to
+        // is_consensus_message_processed
+        #[cfg(debug_assertions)]
+        {
+            let mut seen = HashSet::new();
+            for txn in &txns {
+                assert!(seen.insert(txn.0.key()));
+            }
+        }
+
+        batch.schedule_delete_range(&self.tables.deferred_transactions, &min, &max)?;
+        Ok(txns)
+    }
+
+    // Placeholder implementation
+    fn should_defer(&self, _cert: &VerifiedExecutableTransaction) -> Option<DeferralKey> {
```

Comment:
do we want it to be possible for one tx to have multiple DeferralKeys that are ANDed? or I guess no big deal if we wake up an extra time?

Reply:
I think for now we should just re-defer when necessary. If that becomes a source of wasted work we can revisit